manager/orchestrator/replicatedjob/orchestrator.go

```diff
@@ -45,6 +52,166 @@ func (o *Orchestrator) Run(ctx context.Context) {
+// checking to see if new replicas should be created. reconcileService returns
+// an error if there is some case prevent it from correctly reconciling the
+// service.
+func (o *Orchestrator) reconcileService(id string) error {
+	var (
+		service *api.Service
+		tasks   []*api.Task
+	)
+	// first, get the service and all of its tasks
+	o.store.View(func(tx store.ReadTx) {
+		service = store.GetService(tx, id)
+
+		// getting tasks with FindTasks should only return an error if we've
+		// made a mistake coding; there's no user-input or even reasonable
+		// system state that can cause it. If it returns an error, we'll just
+		// panic and crash.
+		var err error
+		tasks, err = store.FindTasks(tx, store.ByServiceID(id))
+		if err != nil {
+			panic(fmt.Sprintf("error getting tasks: %v", err))
+		}
+	})
+
+	if service == nil {
+		return nil
+	}
+
+	// if this is the first iteration of the service, it may not yet have a
+	// JobStatus, so we should create one if so. this won't actually be
+	// committed, though.
+	if service.JobStatus == nil {
+		service.JobStatus = &api.JobStatus{}
+	}
+
+	// Jobs can be run in multiple iterations. The JobStatus of the service
+	// indicates which Version of iteration we're on. We should only be looking
+	// at tasks of the latest Version
+
+	jobVersion := service.JobStatus.JobIteration.Index
+
+	// now, check how many tasks we need and how many we have running. some of
+	// these tasks may be on their way to completion, but that just means we
+	// won't start enough tasks, which isn't a problem. also, we don't care if
+	// tasks are failed here. failed tasks don't factor into our computation of
+	// how many new tasks to create.
+	//
+	// also, for the math later, we need these values to be of type uint64.
+	runningTasks := uint64(0)
+	completeTasks := uint64(0)
+	// slots keeps track of the slot numbers available for tasks. as long as
+	// the orchestrator isn't broken, then this map will consist of the set of
+	// all integers from 0 to (MaxConcurrent - 1), with a boolean indicating if
+	// that slot is occupied. This slot handling code is much simpler than the
+	// code for replicated services, because we don't need to worry about
+	// restarts.
+	slots := map[uint64]bool{}
+	for _, task := range tasks {
+		// we only care about tasks from this job iteration. tasks from the
+		// previous job iteration are not important
+		// TODO(dperny): we need to stop any running tasks from older job
+		// iterations.
+		if task.JobIteration != nil && task.JobIteration.Index == jobVersion {
+			if task.Status.State == api.TaskStateCompleted {
+				completeTasks++
+			}
+
+			// any tasks that are created and running but not yet terminal are
+			// considered "running tasks" for our purpose, because they don't
+			// need to be created.
+			if task.Status.State <= api.TaskStateRunning && task.DesiredState == api.TaskStateCompleted {
+				runningTasks++
+				slots[task.Slot] = true
```

Comment:
Might be worth checking that `task.Slot` is not in the map yet, and at least log (or return?) an error if it is? Or overkill?

Reply:
Overkill. We control the whole task slot lifecycle. If we have duplicate tasks in a slot, it's our coding bug.

Reply:
Agreed, but still that could be a way to catch said bugs if they appear? Your call :)